pkg/controller/tas/topology_ungater.go

```diff
@@ -396,39 +403,92 @@ func assignGatedPodsToDomainsGreedy(
 func readRanksIfAvailable(log logr.Logger,
 	psa *kueue.PodSetAssignment,
 	pods []*corev1.Pod) (map[int]*corev1.Pod, bool) {
+	if len(pods) == 0 {
+		// If there are no pods then we are done. We do this special check to
+		// ensure we have at least one pod as the code below determines if
+		// rank-ordering is enabled based on the first Pod.
+		return nil, false
+	}
+	if podIndexLabel, rjInfo := determineRanksLookup(pods[0]); podIndexLabel != nil {
+		result, err := readRanksForLabels(psa, pods, *podIndexLabel, rjInfo)
+		if err != nil {
+			log.Error(err, "failed to read rank information from Pods")
+			return nil, false
+		}
+		return result, true
+	}
+	// couldn't determine the labels to lookup the Pod ranks
+	return nil, false
+}
+
+func determineRanksLookup(pod *corev1.Pod) (*string, *replicatedJobsInfo) {
+	// Check if this is JobSet
+	if jobCount, _ := readIntFromLabel(pod, jobset.ReplicatedJobReplicas); jobCount != nil {
+		return ptr.To(batchv1.JobCompletionIndexAnnotation), &replicatedJobsInfo{
+			jobIndexLabel: jobset.JobIndexKey,
+			replicasCount: *jobCount,
+		}
+	}
+	// Check if this is batch/Job
+	if _, found := pod.Labels[batchv1.JobCompletionIndexAnnotation]; found {
+		return ptr.To(batchv1.JobCompletionIndexAnnotation), nil
+	}
+	return nil, nil
+}
+
+func readRanksForLabels(
+	psa *kueue.PodSetAssignment,
+	pods []*corev1.Pod,
+	podIndexLabel string,
+	rjInfo *replicatedJobsInfo,
+) (map[int]*corev1.Pod, error) {
 	result := make(map[int]*corev1.Pod, 0)
-	count := int(*psa.Count)
+	podSetSize := int(*psa.Count)
 	for _, pod := range pods {
-		rank := readIntFromLabel(log, pod, batchv1.JobCompletionIndexAnnotation)
-		if rank == nil {
+		podIndex, err := readIntFromLabel(pod, podIndexLabel)
+		if err != nil {
 			// the Pod has no rank information - ranks cannot be used
-			return nil, false
+			return nil, err
 		}
-		if _, found := result[*rank]; found {
-			// there is a conflict in ranks, they cannot be used
-			return nil, false
+		rank := *podIndex
+		if rjInfo != nil {
+			jobIndex, err := readIntFromLabel(pod, rjInfo.jobIndexLabel)
+			if err != nil {
+				// the Pod has no Job index information - ranks cannot be used
+				return nil, err
+			}
+			singleJobSize := podSetSize / rjInfo.replicasCount
+			if *podIndex >= singleJobSize {
```

Comment:
```suggestion
			if rank >= singleJobSize {
```

Reply:
This would work, but isn't the podIndex better reflect the intention? The rank can still be mutated